### PR TITLE
Prevent No-Folder or root S3 model access

### LINKED ISSUE #1423: [Bug]: Prevent No-Folder (or "root") S3 Model Access
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Today when you're specifying the S3 location for your Model in Model Serving's Deploy Model, you are asked to input a folder path. Submitting empty is `/` (aka the root) -- which once read all the models in the root. Seems this has changed.

### Expected Behavior

We'll want to make the folder path field required (so at least 1 character is required) -- we should prevent this one character from being a `/` so trim then check. 

### Steps To Reproduce

1. Create a Model Server
2. Click on Deploy Modal
3. Fill out the data

### Workaround (if any)

Be more specific with your model location (or model name)

### What browsers are you seeing the problem on?

_No response_

### Open Data Hub Version

```yml
Dashboard: v2.11.0
```

### Anything else

_No response_

### PR DESCRIPTION
**Pull request by manaswinidas**
Fixes #1423 

## Description

1. Renamed "Folder Path" to "Path" 
2. Make "Path" a required field
3. Add helper text to "Path" field
4. Add validation for "Path" field(don't allow `/` or `///////` -> enable "Deploy" button only if "Path" is valid )
Screenshot 2023-07-20 at 2 37 04 AM

GIF(outdated):

Please refer to [this comment](https://github.com/opendatahub-io/odh-dashboard/pull/1512#issuecomment-1650046750) for the latest GIF

https://github.com/opendatahub-io/odh-dashboard/assets/22885912/9754dd01-3955-41fc-a1d5-94fd8dfb39c9

## How Has This Been Tested?

1. Create a Model Server
2. Click on Deploy Modal
3. Fill out the data(Path field in both Existing and New data connection)

## Test Impact

Added tests for the required "Path" field

## Request review criteria:

Self checklist (all need to be checked):
- [X] The developer has manually tested the changes and verified that the changes work
- [X] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [X] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [X] The developer has added tests or explained why testing cannot be added (unit tests & storybook for related changes)

If you have UI changes: 

- [X] Included any necessary screenshots or gifs if it was a UI change.
- [X] Included tags to the UX team if it was a UI/UX change.

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

### CODE REVIEW

**Comment by manaswinidas** [comment]
@vconzola @kywalker-rh @kaedward Is the text "Can either be a folder path or the path to a single model" alright?

Also, as discussed, I haven't added any inline alerts. Just enabling and disabling the "Deploy" button if the "Path" is a valid path as shown in the GIF. 

Here are the other alternatives:

With inline alert and validating tick in case of wrong input:

https://github.com/opendatahub-io/odh-dashboard/assets/22885912/9cc6173f-428d-4e32-8ad4-eb1597b1aac3

With a validating tick(without inline alert):

https://github.com/opendatahub-io/odh-dashboard/assets/22885912/fe3e9a25-952f-4772-a32a-b7eac5be5b4a

Let me know if any of these are a better option in place of the above GIF.

**Review by DaoDaoNoCode — CHANGES_REQUESTED** [review]

**Comment by DaoDaoNoCode** [comment]
BTW, when editing, I can still see the `/` path.

Screenshot 2023-07-18 at 11 24 35 AM

**Comment by vconzola** [comment]
@manaswinidas There are three allowable inputs, correct: (1) a model name in the root folder, (2) a folder path with no model name, or (3) a folder path + model name? If that's true, I prefer more specific helper text without the validation tick and only showing the error helper text when Deploy is selected. I find the helper text and tick changes while typing to be confusing.

I'd like @kaedward to share her opinion, but I think helper text that says something like, "This must be a model name, a folder path, or a folder path and model name." would work. If the input is invalid when the user selects deploy we would show error helper text that says something like, "Not a valid model name or folder path." in red with the error icon displayed in the field as shown here: https://www.patternfly.org/v4/components/helper-text/design-guidelines#when-to-use-helper-text.

**Comment by andrewballantyne** [comment]
> There are three allowable inputs, correct: (1) a model name in the root folder, (2) a folder path with no model name, or (3) a folder path + model name? If that's true...

@vconzola yes that is correct

**Comment by manaswinidas** [comment]
> BTW, when editing, I can still see the `/` path.
> 
> Screenshot 2023-07-18 at 11 24 35 AM

@DaoDaoNoCode Edit seems to work for me, let me know if I'm missing something

https://github.com/opendatahub-io/odh-dashboard/assets/22885912/550b3b19-761f-4681-9d90-c0a33c8436bd

**Comment by DaoDaoNoCode** [comment]
@manaswinidas What if you had a model mounted with the empty path before? Try to edit that one.

**Comment by DaoDaoNoCode** [comment]
Hmm, sounds fair. @andrewballantyne WDYT, do we need to consider the previously created models which could possibly mount on the root dir?

**Comment by andrewballantyne** [comment]
> Hmm, sounds fair. @andrewballantyne WDYT, do we need to consider the previously created models which could possibly mount on the root dir?

I made a [thread in slack](https://redhat-internal.slack.com/archives/C043A9D5DMW/p1689704209069959) -- What is important to note is we need to consider how the flow impacts existing models.

What is the downside from having an "invalid" model deployed against the root `/`? If it locks up the mode from accepting any other inputs, we should address that with UX.

**Comment by kaedward** [comment]
@andrewballantyne For the helper text, how about "Enter the path to a folder or single model."?

**Comment by andrewballantyne** [comment]
> @andrewballantyne For the helper text, how about "Enter the path to a folder or single model."?

@kaedward Yeah, I guess we can go with that -- my only concern is technically I'd view `/` as a folder path heh... Could you sneak an extra couple words in there?

> Enter the path to a non-root folder or single model.

Feels weird still... but it's more technically accurate

**Comment by kaedward** [comment]
@andrewballantyne does this sound any better? 
"Enter a path to a model or folder. This path cannot point to a root folder."

**Comment by andrewballantyne** [comment]
Thanks @kaedward fits my mental model better... @manaswinidas please implement the last text noted by Katie.

> "Enter a path to a model or folder. This path cannot point to a root folder."

**Comment by manaswinidas** [comment]
@andrewballantyne any other changes regarding the error helperText? I can see some comments above: https://github.com/opendatahub-io/odh-dashboard/pull/1512#issuecomment-1640498421

**Comment by andrewballantyne** [comment]
> Also, a suggestion: it will be great if we can check whether the folderPath exists from the backend side(if it's not too far-fetched).

- https://github.com/opendatahub-io/odh-dashboard/issues/1271 🙂 

> @andrewballantyne any other changes regarding the error helperText? I can see some comments above: https://github.com/opendatahub-io/odh-dashboard/pull/1512#issuecomment-1640498421 or we stick to enabling the Deploy button when the input is valid?

@manaswinidas I don't follow what you're asking here -- if you're asking for more wording, pitch the scenario to @kaedward -- if you're asking for what we should do, I think Vince answered that 🤔 Let me know how I can help!

**Comment by manaswinidas** [comment]
>If the input is invalid when the user selects deploy we would show error helper text that says something like, "Not a valid >model name or folder path." in red with the error icon displayed in the field as shown here: >https://www.patternfly.org/v4/components/helper-text/design-guidelines#when-to-use-helper-text.

@andrewballantyne I was just confirming whether there is a need for an error helperText as shown in the comment here. I think I confused Vince with the GIFs 😅

**Comment by kaedward** [comment]
@manaswinidas I think error help text would be useful. Can we detect when an invalid path is entered, or only if the user enters a root folder?

**Comment by manaswinidas** [comment]
We are preventing inputs like `/` or `/////` It needs to be some text that can contain `/` but not only `/`. As of now, I'm enabling the `Deploy` button if the `Path` is valid. Here's a GIF for your reference. Does this look good or it still needs an error helperText?

https://github.com/opendatahub-io/odh-dashboard/assets/22885912/e43c2e5b-9bea-4aa4-9bad-b7cb9bb99c6d

**Comment by manaswinidas** [comment]
Also, @kaedward @andrewballantyne is there a need to mention the link to the docs for why we are preventing the root folder path? Here is the link to the docs: https://docs.openvino.ai/2023.0/ovms_docs_models_repository.html

**Comment by kaedward** [comment]
@manaswinidas If it's not too much trouble, I think adding an error helper text would be helpful, even if it is only for when a root folder is entered. Maybe "The path must not point to a root folder."

I'm not sure the doc link is necessary here, but I'm open to including it.

**Comment by andrewballantyne** [comment]
Let us not link to any docs -- if they break, we are on the hook to fixing it. We can document this through product documentation downstream and in the new website upstream -- we don't need to address this at the code level.

Works for me Katie -- on the error when root folder. cc @vconzola FYI we are shifting again on this.

**Comment by vconzola** [comment]
@kaedward I don't understand where/when you're suggesting that error be displayed?

**Comment by manaswinidas** [comment]
We have 2(or 3) options:
1. Show the error message while the user is typing the input. Do we still disable the "Deploy" button when the error is getting displayed? Or enable the "Deploy" button and make the user try to send the wrong input, and after they click on "Deploy"- show them the error. 
(My concerns in the first case: Changing the helperText to the error helperText while typing might confuse the user coz it draws attention

My concerns for the second case(disputed opinion): I've found it annoying a user when I get an error after clicking on the "Submit" button when the form/field could have shown the error while doing it, thus saving me time.)

2. Disable the "Deploy" button when the user types `/` or `/////` and enable it when the user types the correct input. 

Would you like me to send GIFs for some or all of these scenarios?

**Comment by andrewballantyne** [comment]
> @kaedward I don't understand where/when you're suggesting that error be displayed?

@vconzola @manaswinidas @kaedward okay, let me step in here and cap this off. 

My suggested state of this modal is this, let me know if you agree.

For the folder field...
1. Empty initial state => show the empty field, deploy button is disabled (invalid input)
    * the red asterisk will help the user know the field is needed
    * the help text of "Enter a path to a model or folder. This path cannot point to the root folder."
3. If the user types `/` we show them the error under the field "The path must not point to a root folder."
    * The error goes away (and the button enables -- if the rest of the modal is good) when they continue `/<any character>` -- eg `/a`
    * We stay in error if they type more than 1 `/` -- `///////` scenario; even if they type `////<any character>`
    * We show the error using PF field validation ([link](https://www.patternfly.org/v4/components/form#invalid))
4. If the user straight up types a non slash character, we don't show anything in the field but we enable the button (if the rest of the modal is good)
    * I think it's important we don't use success validation here because we are not validating the validity of the path

Does that make sense to everyone? If not, let us take this to slack (or a meeting), and it'll likely need more back and forth.

**Comment by vconzola** [comment]
@andrewballantyne @manaswinidas @kaedward I'm OK with everything @andrewballantyne suggested with one exception. i think it's going to look weird if we immediately show the error as soon as the user types"/". It could cause the error message to flash briefly if they don't type the next, non-/ character quickly enough. Can we add a short (500 ms to 1 s - will probably need to play around with it to see what feels right) delay before displaying the error below the field?

**Comment by manaswinidas** [comment]
@kaedward @vconzola Do we have a success helperText too or just the default helperText "Enter a path to a model or folder. This path cannot point to the root folder." will do?

**Comment by vconzola** [comment]
@manaswinidas We don't need any success helperText - only if there's an error.

**Comment by manaswinidas** [comment]
@vconzola @kaedward Final GIF:

https://github.com/opendatahub-io/odh-dashboard/assets/22885912/b8637985-b489-4a67-b107-e3cbf6c537a3

**Comment by vconzola** [comment]
lgtm

File: frontend/src/pages/modelServing/screens/projects/InferenceServiceModal/DataConnectionFolderPathField.tsx [hunk 1]
```diff
@@ -9,19 +10,47 @@ type DataConnectionFolderPathFieldProps = {
...
-        type="text"
-        value={folderPath}
-        placeholder="eg. data"
-        onChange={(value) => setFolderPath(value)}
-      />
-    </InputGroup>
-  </FormGroup>
-);
+}) => {
+  type validate = 'success' | 'warning' | 'error' | 'default';
```
Comment (andrewballantyne) [inline frontend/src/pages/modelServing/screens/projects/InferenceServiceModal/DataConnectionFolderPathField.tsx:14]: Types are always PascalCase

Also, why do you need this? Infer the data from PF. 

```
React.ComponentProps<typeof FormGroup>['validated']
```

That should get you the value behind the prop of FormGroup's prop `validated`.

**Comment by manaswinidas** [comment]
> For what it is worth -- don't put success intentionally. We are not validating the value, we are just helping them avoid known failure points. Success would mislead the user to think we know if their folder path is accurate -- which we do not.

I was reiterating the same yesterday, we don't have a success helperText and the check mark in the textbox for the same reason. Also, I've added a 1s delay, let me know if that needs to be changed. [This comment](https://github.com/opendatahub-io/odh-dashboard/pull/1512#issuecomment-1650046750) has the latest GIF.

File: frontend/src/pages/modelServing/screens/projects/InferenceServiceModal/DataConnectionFolderPathField.tsx [hunk 2]
```diff
@@ -9,19 +10,47 @@ type DataConnectionFolderPathFieldProps = {
...
+
+  const handlePathChange = (folderPath: string) => {
+    setFolderPath(folderPath);
+    setValidated('default');
+  };
+  React.useEffect(() => {
+    const timer = setTimeout(() => {
+      if (folderPath === '') {
+        setValidated('default');
+      } else if (folderPath === '/' || folderPath.includes('//')) {
```
Comment (andrewballantyne) [inline frontend/src/pages/modelServing/screens/projects/InferenceServiceModal/DataConnectionFolderPathField.tsx:26]: If you get into an error -- I think you're going to not get out of it on `paste`... can you switch your logic around `if` this, `else` set back to default?

**Review by andrewballantyne — CHANGES_REQUESTED** [review]

**Review by andrewballantyne — COMMENTED** [review]
Thanks for making the changes -- you really should use PascalCase for your Type variables... but this solves the real problem I saw. 

/approve

Give me a moment to spin up my UI to test it, but should work fine.

**Review by lucferbux — APPROVED** [review]
/lgtm

**Comment by manaswinidas** [comment]
Issue logged for changes discussed in https://github.com/opendatahub-io/odh-dashboard/pull/1512#issuecomment-1652037627 : https://github.com/opendatahub-io/odh-dashboard/issues/1601

**Pull request merged**